File: leetcode/118/PascalTriangle.java
import java.util.*;

class PascalTriangle {
    public static void main(String[] args) {
        Solution sl = new Solution();
        System.out.println(sl.generate(5));
    }
}

class Solution {
    public List<List<Integer>> generate(int numRows) {

        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> prevList = new ArrayList<Integer>();

        for (int r = 1; r <= numRows; r++) {
            List<Integer> tempList = new ArrayList<Integer>();
            for (int c = 1; c <= r; c++) {
                if (c == 1 || c == r) {
                    tempList.add(1);
                } else {
                    tempList.add(prevList.get(c - 2) + prevList.get(c - 1));
                }
            }

            prevList = tempList;
            result.add(tempList);
        }

        return result;

    }
}

// class Solution {
// public List<List<Integer>> generate(int numRows) {
// List<List<Integer>> result = new ArrayList<List<Integer>>();
// for (int r = 1; r <= numRows; r++) {
// List<Integer> tempList = new ArrayList<Integer>();
// for (int c = 1; c <= r; c++) {
// if (c == 1 || c == r) {
// tempList.add(1);
// } else {
// System.out.println(r + "," + c);
// System.out.println(result);
// tempList.add(result.get(r - 2).get(c - 2) + result.get(r - 2).get(c - 1));
// }
// }
// result.add(tempList);
// }
// return result;
// }
// }

class Solution {
    public List<List<Integer>> generate(int numRows) {

        List<List<Integer>> result = new ArrayList<List<Integer>>();

        for (int r = 0; r < numRows; r++) {
            List<Integer> tempList = new ArrayList<Integer>();
            for (int c = 0; c < r + 1; c++) {
                if (c == 0 || c == r) {
                    tempList.add(1);
                } else {
                    // System.out.println(r + "," + c);
                    // System.out.println(result);
                    tempList.add(result.get(r - 1).get(c - 1) + result.get(r - 1).get(c));
                }
            }
            result.add(tempList);
        }

        return result;

    }
}

File: leetcode/118/PascalTriangle.js
/**
 * @param {number} numRows
 * @return {number[][]}
 */
// let generate = function(numRows) {
//     let final=[];
//     for(let i=0;i<numRows;i++){
//         final[i]=[];
//         final[i][0]=final[i][i]=1;
//         for(let j=1;j<i;j++){
//             final[i][j]=final[i-1][j]+final[i-1][j-1];
//         }
//     }
//     return final;
// };

// var generate = function(numRows) {
//     let ret = [];
    
//     for(let i = 0; i < numRows; i++)
//     {
//         ret[i] = [];
//         ret[i][0] = ret[i][i] = 1;
//         for(let j = 1; j < i; j++)
//         {
//                ret[i][j] = ret[i-1][j] + ret[i-1][j-1]; 
//         }
//     }
//     return ret;
// };

/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function(numRows) {
    let result = [];
    
    for(let i = 0; i < numRows; i++)
    {  
        result[i] = [];
        for(let j = 0; j < i+1; j++)
        {
            if(j===0||j=== i)
            {
                result[i][j] = 1;
            }else
            {
                console.log(i,j,i-1,j-1);
                
                result[i][j] = result[i-1][j-1] + result[i-1][j];
                console.log(result[i][j]);
            }
        }
    }
    return result;
};

console.log(generate(5));